### PR TITLE
Add support for GCC/MinGW compilers

### PR DESCRIPTION
**Pull request by Mixaill**
* Fixes compilation under GCC.
* Fixes AVX2 detection for Clang

Tested under 
* Ubuntu 19.04 AMD64
   * GCC 8.3
   * Clang 8.0
* Windows 10 AMD64
   * MSVC 16.3 Preview 1
   * MinGW 9.1

Test suite is OK for all cases: 476/476 (https://github.com/Mixaill/directxmathtest/tree/gcc-fixes)

### CODE REVIEW

**Comment by walbourn** [comment]
Thanks for the reminder to update all the auxiliary stuff in the repo. You should merge the latest changes to your fork:

[XDSP](https://github.com/microsoft/DirectXMath/commit/0aa6a41e8f7ef3b245e518a389c7b8f2ccdba27b
)

**Comment by Mixaill** [comment]
@walbourn rebased and cleaned up

**Comment by walbourn** [comment]
Sorry, wasn't quite done...

[SHMath](https://github.com/microsoft/DirectXMath/commit/dd0a88f83f83d852e6fcd75173be6c577f95835a)

One more coming shortly for the 'Extensions'

File: XDSP/XDSP.h [hunk 1]
```diff
@@ -676,7 +676,11 @@ namespace XDSP
         }
         else
         {
+            #ifdef __GNUC__
+            memcpy(vRealTemp, pReal, (uLength>>2)*sizeof(XMVECTOR));
+            #else
             memcpy_s(vRealTemp, sizeof(vRealTemp), pReal, (uLength>>2)*sizeof(XMVECTOR));
```
Comment (walbourn) [inline XDSP/XDSP.h]: ``memcpy_s`` is part of the C11 standard. Shouldn't it be supported?

In any case, you can probably fix this more easily with a ``#define memcpy_s(d,ds,s,c) memcpy(d,s,c)

File: Inc/DirectXMathMisc.inl [hunk 2]
```diff
@@ -2026,8 +2026,8 @@ inline bool XMVerifyCPUSupport()
         return false; // No SSE2/SSE support
 
 #if defined(__AVX2__) || defined(_XM_AVX2_INTRINSICS_)
-#ifdef __clang__
-    __cpuid_count(0, 7, CPUInfo[0], CPUInfo[1], CPUInfo[2], CPUInfo[3]);
+#if defined(__clang__) || defined(__GNUC__) 
+    __cpuid_count(7, 0, CPUInfo[0], CPUInfo[1], CPUInfo[2], CPUInfo[3]);
```
Comment (walbourn) [inline Inc/DirectXMathMisc.inl]: Good catch. Fixed in [this commit](https://github.com/microsoft/DirectXMath/commit/549b51d9cc328903739d5fbe7f348aec00c83339)

File: Inc/DirectXMath.h [hunk 3]
```diff
@@ -38,60 +40,80 @@
...
+#if !defined(_XM_FMA3_INTRINSICS_) && (defined(_XM_AVX2_INTRINSICS_) && !defined(__GNUC__)) && !defined(_XM_NO_INTRINSICS_)
+#define _XM_FMA3_INTRINSICS_
+#endif
+
+#if !defined(_XM_FMA3_INTRINSICS_) && defined(__FMA__) && !defined(_XM_NO_INTRINSICS_)
 #define _XM_FMA3_INTRINSICS_
 #endif
 
-#if !defined(_XM_F16C_INTRINSICS_) && defined(_XM_AVX2_INTRINSICS_) && !defined(_XM_NO_INTRINSICS_)
+#if !defined(_XM_F16C_INTRINSICS_) && (defined(_XM_AVX2_INTRINSICS_) && !defined(__GNUC__)) && !defined(_XM_NO_INTRINSICS_)
```
Comment (walbourn) [inline Inc/DirectXMath.h]: Why did you add all these !__GNUC__ guards? These shouldn't be enabled without explicit opt-in in any case.

Comment (Mixaill) [reply Inc/DirectXMath.h]: Well:
* User set up `-mavx2` compiler flag, it defines `__AVX2__`
* `__AVX2__` defines `_XM_AVX2_INTRINSICS_`
* `_XM_AVX2_INTRINSICS_` defines `_XM_F16C_INTRINSICS_`
* and here we got compilation error, because for F16C instrinsics we should explicitly set up `-mf16c` compile flag.

Comment (Mixaill) [reply Inc/DirectXMath.h]: So, that's why I switched to use compiler definitions for GCC, it much safer.

File: XDSP/XDSP.h (hunk 1 above)
Comment (Mixaill) [reply XDSP/XDSP.h]: Still missing in G++ :(

**Comment by Mixaill** [comment]
@walbourn rebased and cleaned up. For `#ifdef __GNUC__` read previous comment.

**Comment by walbourn** [comment]
Sorry, took longer than I expected. Extensions updated in this [commit](https://github.com/microsoft/DirectXMath/commit/24a4887b35b34f3bf346730e4920644fc2b8dd6a).

**Comment by Mixaill** [comment]
@walbourn rebased

File: Inc/DirectXMath.h (hunk 3 above)
Comment (walbourn) [reply Inc/DirectXMath.h]: So this is "By Design". If you use AVX2, it assumes you can also use F16C and FMA because they are inclusive. The rules follow all the way down.  In fact, with clang setting ``-mavx2`` implies ``-mf16c``.

File: Extensions/DirectXMathAVX.h [hunk 4]
```diff
@@ -28,7 +28,7 @@ inline bool XMVerifyAVXSupport()
 
     // See http://msdn.microsoft.com/en-us/library/hskdteyh.aspx
     int CPUInfo[4] = {-1};
-#ifdef __clang__
+#if defined(__clang__) || defined(__GNUC__)
```
Comment (walbourn) [inline Extensions/DirectXMathAVX.h]: From what I understand clang/LLVM also defines ``__GNUC__`` so these could all be simplified to:

```
#ifdef __GNUC__
```

Correct?

File: Inc/DirectXMath.h [hunk 5]
```diff
@@ -1905,8 +1956,12 @@ template<uint32_t VSLeftRotateElements, uint32_t Select0, uint32_t Select1, uint
 // separate math routine it would be reloaded.
 
 #ifndef XMGLOBALCONST
```
Comment (walbourn) [inline Inc/DirectXMath.h]: What's the reason for the additional guard here? Doesn't ``attribute(weak)`` work in all cases?

File: Inc/DirectXMath.h [hunk 6]
```diff
@@ -141,7 +163,27 @@
 #endif
 #endif // !_XM_NO_INTRINSICS_
 
+#ifdef _WIN32
 #include <sal.h>
+#else
+#define _Analysis_assume_(x)
```
Comment (walbourn) [inline Inc/DirectXMath.h]: My general answer is to have a sal.h in your system paths that does this... Avoids cluttering up the header.

File: Inc/DirectXMath.h (hunk 3 above)
Comment (walbourn) [reply Inc/DirectXMath.h]: and since ``__GNUC__`` exists for clang/LLVM, this is likely to break it for clang. Just document that for GNUC you need to use ``-mavx2 -mf16c -mfma`` is safer and avoids breaking clang/LLVM

File: Inc/DirectXMath.h (hunk 6 above)
Comment (Mixaill) [reply Inc/DirectXMath.h]: You have no possibility to have sal.h on Linux.

File: Inc/DirectXMath.h (hunk 5 above)
Comment (Mixaill) [reply Inc/DirectXMath.h]: Fails on MinGW, which for same reason ahve support for `__declspec(selectany)`

File: Extensions/DirectXMathAVX.h (hunk 4 above)
Comment (Mixaill) [reply Extensions/DirectXMathAVX.h]: I'm not sure about Clang/C2 compiler, looks like it should not have such definition. (By the way, is it supported by DirectXMath at all?)

Comment (walbourn) [reply Extensions/DirectXMathAVX.h]: Clang/C2 was experimental and is now abanonded, and it never implemented any of the SSE intrinsics. It 'kinda' worked with ``_XM_NO_INTRINSICS_`` but Clang/C2 was mostly about improving conformance of source.

File: Inc/DirectXMath.h (hunk 3 above)
Comment (Mixaill) [reply Inc/DirectXMath.h]: OK, removed

File: Extensions/DirectXMathAVX.h (hunk 4 above)
Comment (Mixaill) [reply Extensions/DirectXMathAVX.h]: Well, looks like Clang on Windows with MSVC flavour (`clang-cl.exe`) (at least, which is builtin with MSVS 2019 16.3) does not defines `__GNUC__`. 

So we should use both definitions for CPUID.

**Comment by Mixaill** [comment]
@walbourn cleaned up and retested with:
* MSVC 16.3, MinGW 9.1 and Clang 8.0 on Windows 10 
* GCC 9.1 and Clang 8.0 on Ubuntu 19.10

**Comment by AndySomogyi** [comment]
Hi, just found this PR. I'm trying to use DirectXMath on Linux / Mac, I'd be happy to test on Mac.

File: Inc/DirectXMath.h (hunk 6 above)
Comment (walbourn) [reply Inc/DirectXMath.h]: Actually, there's an open source version of SAL.H under the MIT license you can use [here](https://github.com/dotnet/corert/blob/master/src/Native/inc/unix/sal.h)

Comment (AndySomogyi) [reply Inc/DirectXMath.h]: Can you include the open source sal.h, and have an ifdef, something like

```
#ifndef MSVC
#include "opensoruce_sal.h"
#else
#include "sal.h"
#endif
```

???

Comment (Mixaill) [reply Inc/DirectXMath.h]: Yes, but it is difficult to tell users that they should take header file somewhere from another project. 

Ideally, the library without dependencies should be without dependencies :)

Comment (AndySomogyi) [reply Inc/DirectXMath.h]: Updated my prev comment, I guess it wasn't super clear. I meant to say package the open source sal.h with the DirectXMath repo, and use it if we're not compiling with MSVC.

Comment (walbourn) [reply Inc/DirectXMath.h]: sal.h is already there for clang for Windows as well.

Comment (AndySomogyi) [reply Inc/DirectXMath.h]: Picks up the sal.h, but the compiler is not super happy...

```
/Users/andy/src/mechanica/extern/DirectXMath/Inc/DirectXMath.h:1880:25: warning: calling convention '__fastcall' ignored for this target [-Wignored-attributes]
    inline XMVECTOR     XM_CALLCONV     XMVectorRotateRight(FXMVECTOR V)
                        ^
/Users/andy/src/mechanica/extern/DirectXMath/Inc/DirectXMath.h:29:21: note: expanded from macro 'XM_CALLCONV'
#define XM_CALLCONV __fastcall
                    ^
/Users/andy/src/mechanica/extern/DirectXMath/Inc/DirectXMath.h:1887:25: warning: calling convention '__fastcall' ignored for this target [-Wignored-attributes]
    inline XMVECTOR     XM_CALLCONV     XMVectorInsert(FXMVECTOR VD, FXMVECTOR VS)
                        ^
/Users/andy/src/mechanica/extern/DirectXMath/Inc/DirectXMath.h:29:21: note: expanded from macro 'XM_CALLCONV'
#define XM_CALLCONV __fastcall
                    ^
/Users/andy/src/mechanica/extern/DirectXMath/Inc/DirectXMath.h:1911:1: error: '__declspec' attributes are not enabled; use '-fdeclspec' or '-fms-extensions' to enable support for __declspec attributes
XMGLOBALCONST XMVECTORF32 g_XMSinCoefficients0      = { { { -0.16666667f, +0.0083333310f, -0.00019840874f, +2.7525562e-06f
```

Comment (AndySomogyi) [reply Inc/DirectXMath.h]: Looking ad adding the '-fdeclspec' or '-fms-extensions' options now...

Comment (walbourn) [reply Inc/DirectXMath.h]: Adding ``-fms-extensions`` will fix the issues for x64/x86.

For ARM/ARM64, you need some additional #defines for missing intrinsics because all the ``_ex`` versions which take an alignment are unsupported on non-Windows ARM platforms. These can be easily mapped to their non-ex versions.

Comment (AndySomogyi) [reply Inc/DirectXMath.h]: OK, compiles, still has tons of ```warning: calling convention '__fastcall' ignored for this target``` warnings. 

Looking into this now...

Comment (Mixaill) [reply Inc/DirectXMath.h]: The easiest way to get rid off  these warnings is to replace

```
#elif !defined(__GNUC__) 
#define XM_CALLCONV __fastcall
```

with

```
#elif !defined(__GNUC__) && !defined(__APPLE__ )
#define XM_CALLCONV __fastcall
```

Comment (AndySomogyi) [reply Inc/DirectXMath.h]: What *seems* to work fine, is :

```
#if defined(__APPLE__) 
#define __fastcall
#endif
#include "DirectXMath.h"
#endif
```

No warnings, lets test some functionality....

**Comment by ahmetabdi** [comment]
Same any progress on this?

**Comment by Mixaill** [comment]
Well, looks like currently we have several unsolved issues:
- untested on MacOs
- "sal.h" on non-Windows platforms.

**Comment by Mixaill** [comment]
Rebased and retested:
* Ubuntu 20.04 amd64
  - GCC 9.2.1-30ubuntu2: 476/476
  - Clang 9.0.1-8ubuntu1: 476/476

File: Inc/DirectXMath.h [hunk 7]
```diff
@@ -76,14 +82,14 @@
 #if !defined(_XM_ARM_NEON_INTRINSICS_) && !defined(_XM_SSE_INTRINSICS_) && !defined(_XM_NO_INTRINSICS_)
 #if (defined(_M_IX86) || defined(_M_X64) || __i386__ || __x86_64__) && !defined(_M_HYBRID_X86_ARM64)
 #define _XM_SSE_INTRINSICS_
-#elif defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || __arm__ || __aarch64__
+#elif (defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || __arm__ || __aarch64__) && !defined(__GNUC__)
```
Comment (walbourn) [inline Inc/DirectXMath.h]: Actually, I think GNU's ARM-NEON should work. I just needed to guard the few MSVC extensions... I'm doing that now.

**Comment by walbourn** [comment]
I've add 90% of your changes including fixes to the ARM-NEON path to build with clang / GCC in [this pull request](https://github.com/microsoft/DirectXMath/pull/102).

**Comment by Mixaill** [comment]
@walbourn I think it would be useful to add a little instruction to `README.me` where to download the `sal.h` file.

**Comment by walbourn** [comment]
Agreed. I mentioned it in the [wiki](https://github.com/microsoft/DirectXMath/wiki#compiler-support), but I can add it to the readme as well.

**Comment by walbourn** [comment]
Updated readme in [this commit](https://github.com/microsoft/DirectXMath/commit/438779204397eb42a74606c81927e0b92843a84a)
